tensorpipe/transport/shm/reactor.cc

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/transport/shm/reactor.h>
+
+#include <tensorpipe/util/ringbuffer/shm.h>
+
+namespace tensorpipe {
+namespace transport {
+namespace shm {
+
+Reactor::Reactor() {
+  int headerFd;
+  int dataFd;
+  std::shared_ptr<TReactorRingBuffer> rb;
+  std::tie(headerFd, dataFd, rb) =
+      util::ringbuffer::shm::create<TReactorRingBuffer>(kSize);
+  headerFd_ = Fd(headerFd);
+  dataFd_ = Fd(dataFd);
+  consumer_.emplace(rb);
+  producer_.emplace(rb);
+  thread_ = std::thread(&Reactor::run, this);
+}
+
+Reactor::~Reactor() {
+  done_.store(true);
+  thread_.join();
+}
+
+Reactor::TToken Reactor::add(TFunction fn) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  TToken token;
+
+  // Either reuse a token or generate a new one.
+  auto it = reusableTokens_.begin();
+  if (it != reusableTokens_.end()) {
+    token = *it;
+    reusableTokens_.erase(it);
+  } else {
+    token = nextToken_++;
+  }
+
+  // Ensure there is enough space in the functions vector.
+  if (functions_.size() <= token) {
+    functions_.resize(token + 1);
+  }
+
+  functions_[token] = std::move(fn);
+  return token;
+}
+
+void Reactor::remove(TToken token) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  functions_[token] = nullptr;
+  reusableTokens_.insert(token);
+}
+
+void Reactor::trigger(TToken token) {
+  for (;;) {
+    auto rv = producer_->write(token);
+    if (rv == -EAGAIN) {
+      continue;
+    }
+    TP_DCHECK_EQ(rv, sizeof(token));
+    break;
+  }
+}
+
+std::tuple<int, int> Reactor::fds() const {
+  return std::make_tuple(headerFd_.fd(), dataFd_.fd());
+}
+
+void Reactor::run() {
+  while (!done_.load()) {
+    uint32_t token;
+    auto ret = consumer_->copy(token);
+    if (-ret == ENODATA) {
```

Comment:
Nit: `ret == -ENODATA`. It's consistent with everything else and I believe this way the compiler can do the negation at compile time rather than at runtime.